presto-google-sheets/src/main/java/io/prestosql/plugin/google/sheets/GSheetsDataProvider.java

```diff
@@ -0,0 +1,227 @@
+            throw new PrestoException(SHEETS_BAD_CREDENTIALS_ERROR, e);
+        }
+        OptionalLong expiresAfterWriteMillis = OptionalLong.of(config.getSheetsDataExpireAfterWrite().toMillis());
+        long maxCacheSize = config.getSheetsDataMaxCacheSize();
+        ExecutorService executor = newDirectExecutorService();
+
+        this.tableSheetMappingCache = newCacheBuilder(expiresAfterWriteMillis, maxCacheSize)
+                .build(asyncReloading(new CacheLoader<String, Optional<String>>() {
+                    @Override
+                    public Optional<String> load(String tableName)
+                            throws Exception
+                    {
+                        return getSheetExpressionForTable(tableName);
+                    }
+
+                    @Override
+                    public Map<String, Optional<String>> loadAll(Iterable<? extends String> tableList)
+                            throws Exception
+                    {
+                        return getSheetExpressionForAllTables();
+                    }
+                }, executor));
+
+        this.sheetDataCache = newCacheBuilder(expiresAfterWriteMillis, maxCacheSize)
+                .build(asyncReloading(from(this::getDataForSheetExpression), executor));
+    }
+
+    private Optional<String> getSheetExpressionForTable(String tableName)
+    {
+        try {
+            List<List<Object>> data = readAllValuesFromSheetExpression(metadataSheetId);
+            // first line is assumed to be sheet header
+            if (data.size() > 1) {
+                for (int i = 1; i < data.size(); i++) {
+                    if (data.get(i).size() >= 2) {
+                        String tableId = String.valueOf(data.get(i).get(0));
+                        String sheetId = String.valueOf(data.get(i).get(1));
+                        if (tableName.equalsIgnoreCase(tableId)) {
+                            return Optional.of(sheetId);
+                        }
+                    }
+                }
+            }
+        }
+        catch (Exception e) {
+            throw new PrestoException(SHEETS_METASTORE_ERROR, e);
+        }
+        return Optional.empty();
+    }
+
+    private Map<String, Optional<String>> getSheetExpressionForAllTables()
+    {
+        Map<String, Optional<String>> tableSheetMap = new HashMap<>();
+        try {
+            List<List<Object>> data = readAllValuesFromSheetExpression(metadataSheetId);
+            // first line is assumed to be sheet header
+            if (data.size() > 1) {
+                for (int i = 1; i < data.size(); i++) {
+                    if (data.get(i).size() >= 2) {
+                        String tableId = String.valueOf(data.get(i).get(0));
+                        String sheetId = String.valueOf(data.get(i).get(1));
+                        tableSheetMap.put(tableId, Optional.of(sheetId));
+                    }
+                }
+            }
+        }
+        catch (Exception e) {
```

Comment:
Throwing and catching plain `Exception` is too broad. Use more dedicated and more meaningful exceptions.